Source/ParticleInterpolators/ParticleInterpolators.impl.hpp

```diff
@@ -0,0 +1,655 @@
+            });
+
+        amrex::Gpu::streamSynchronize();
+    }
+
+    // amrex::ParallelDescriptor::Barrier(); // TODO! test if this makes a
+    // difference.
+}
+
+// interpolate variables into SOA slots
+template <int num_components>
+void ParticleInterpolators<num_components>::interpolate_to_particle()
+{
+    AMREX_ASSERT(m_initialized);
+
+    ensure_redistributed();
+
+    const int start_comp = m_start_comp;
+    const int ncomp      = num_components;
+
+    for (int lev = 0; lev <= m_gr_amr->finestLevel(); ++lev)
+    {
+        // if particles are only on the coarsest level, skip the looping over
+        // levels
+        if (this->NumberOfParticlesAtLevel(lev) == 0)
+            continue;
+
+        amrex::AmrLevel &level      = m_gr_amr->getLevel(lev);
+        const amrex::Geometry &geom = level.Geom();
+        amrex::MultiFab &state      = level.get_new_data(static_cast<int>(0));
+
+        AMREX_ASSERT(start_comp + ncomp <= state.nComp());
+
+        amrex::IntVect nghost(AMREX_D_DECL(2, 2, 2));
+        state.FillBoundary(start_comp, ncomp, nghost, geom.periodicity());
+
+        const auto plo = geom.ProbLoArray();
+        const auto dxi = geom.InvCellSizeArray();
+
+        // loop over tiles and interpolate now
+        for (ParIterType it(*this, lev); it.isValid(); ++it)
+        {
+            auto &ptile    = this->ParticlesAt(lev, it);
+            auto ptd       = ptile.getParticleTileData();
+            const int np   = it.numParticles();
+            auto fab_array = state[it].const_array();
+
+            amrex::ParallelFor(
+                np,
+                [=] AMREX_GPU_DEVICE(int ip)
+                {
+                    auto &sp = ptd[ip];
+
+                    amrex::IntVect is_nodal = amrex::IntVect::TheZeroVector();
+                    LagrangeInterpolator<5> interp; // 4th order interpolation
+                    interp.compute_weights(sp, plo, dxi, is_nodal);
+
+                    amrex::ParticleReal vals[AMREX_SPACEDIM];
+                    interp.interpolate(&fab_array, vals, start_comp, ncomp);
+
+                    // write results to SOA
+                    for (int k = 0; k < ncomp; ++k)
+                    {
+                        ptd.rdata(k)[ip] = vals[k];
+                    }
+                });
+        }
+
+        // synchronize GPU streams to ensure all particles are updated
+        amrex::Gpu::streamSynchronize();
+    }
+
+    m_particles_seeded  = true;
+    m_need_redistribute = false;
+}
+
+// Interpolation for derived vars, takes in MultiFab and comps (unique
+// components numbers), as e.g. we may want to interpolate several fields at
+// once. However, as per current implementation, we can have only contiguous
+// comps
+template <int num_components>
+void ParticleInterpolators<num_components>::
+    interpolate_to_particle_from_derived_fields(
+        const std::vector<const amrex::MultiFab *> &fields_by_lev)
```

Comment:
Does the `amrex::MultiFab*` type need to be `const`?

I think we should change the variable name as well hence my suggest below.
```suggestion
        const std::vector<amrex::MultiFab *> &a_derived_mf_vect)
```